PySpark DataFrame API Support

eakmanrq opened the pull request:
This contains a drop-in replacement for the PySpark DataFrame API (see README for details).

tobymao commented on dev-requirements.txt:
```diff
@@ -0,0 +1,2 @@
+-r requirements.txt
```
requirements.txt is dev-requirements, add the rest into here and delete the old file

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+
+
+class Column:
+    def __init__(self, expression: t.Union[t.Any, exp.Expression]):
```
should you use maybe_parse?

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+    @property
+    def column_expression(self) -> exp.Column:
+        if self.is_alias:
+            return self.expression.args["this"]
```
expression.this

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+        return self.expression
+
+    @property
+    def alias_or_name(self):
```
missing return type

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+    def alias_or_name(self):
+        if isinstance(self.expression, exp.Null):
+            return "NULL"
+        if isinstance(self.expression.args.get("this"), exp.Star):
```
replace all places where you do args["this"] to expression.this

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+        return self.expression.sql(dialect="spark", **kwargs)
+
+    def alias(self, name: str) -> Column:
+        new_expression = exp.Alias(alias=exp.to_identifier(name, quoted=True), this=self.column_expression)
```
can you use exp.alias_ instead?

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+
+        true_value = value if isinstance(value, Column) else lit(value)
+        new_column = self.copy()
+        new_column.expression.args["default"] = true_value.column_expression
```
never set args directly, use set

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+    def over(self, window: WindowSpec) -> Column:
+        window_args = window.expression.args
+        window_args["this"] = self.column_expression
+        return Column(exp.Window(**window_args))
```
can you add all places where you instantiate an expression to have kwargs, like this=

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,741 @@
+
+    @property
+    def latest_cte_name(self) -> str:
+        if len(self.expression.ctes) == 0:
```
if not self.expression.ctes

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,741 @@
+        ctes = expression.ctes
+        replacement_mapping = {}
+        for cte in ctes:
+            old_name_id = cte.args["alias"].this
```
can you just do cte.alias?

eakmanrq replied:
I need the id value and it looks like `cte.alias` returns a string

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,741 @@
+            new_hashed_id = exp.to_identifier(
+                self._create_hash_from_expression(cte.this), quoted=old_name_id.args["quoted"]
+            )
+            replacement_mapping[old_name_id] = new_hashed_id
```
there's a helper method exp.alias_(table=True) you can use

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,741 @@
+        new_expression = df._add_ctes_to_expression(new_expression, expression.ctes + [cte_expression])
+        sel_columns = df._get_outer_select_columns(cte_expression)
+        star_columns = [col for col in sel_columns if isinstance(col.expression.this, exp.Star)]
+        if len(star_columns) > 0:
```
if star_columns

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,741 @@
+
+    def _hint(self, hint_name: str, args: t.List[Column]) -> DataFrame:
+        hint_name = hint_name.upper()
+        if hint_name in {
```
make this a constant?

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,741 @@
+            filtered_ctes = [
+                cte
+                for cte in select_expression.ctes
+                if cte.alias_or_name not in {x.alias_or_name for x in replacement_mapping.values()}
```
replaceme_mapping double for loop?

eakmanrq replied:
Yeah this is part of the hacky section that will be replaced once we improve the optimizer to remove unused CTEs.

eakmanrq commented on sqlglot/dataframe/README.md:
```diff
@@ -0,0 +1,224 @@
+    .agg(F.countDistinct(F.col("employee_id")).alias("num_employees")) 
+)
+
+print(df.sql())  # Spark will be the dialect used by default
```
It is pretty by default: https://github.com/tobymao/sqlglot/blob/5f2b954a9e940676cdb8512a707a95d8b40c23a3/sqlglot/dataframe/sql/dataframe.py#L318

eakmanrq replied:
This feels more to me like a high level usage of sqlglot and therefore pretty by default instead of low level calls where we assume not pretty.

tobymao replied:
all sqglot should be non pretty by default, users can set pretty=True if they want the other behavior

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,741 @@
+                        cte
+                        for cte in self.expression.ctes
+                        if cte.alias_or_name in cte_names_in_join
+                        and ambiguous_col.alias_or_name in cte.args["this"].named_selects
```
cte.this

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,741 @@
+
+        columns = self._ensure_and_normalize_cols(subset) if subset else all_columns
+        if isinstance(to_replace, dict):
+            old_values = list(to_replace.keys())
```
list(to_replace) don',t need keys

tobymao commented on sqlglot/dataframe/sql/functions.py:
```diff
@@ -0,0 +1,1251 @@
+
+def lit(value: t.Optional[t.Any] = None) -> Column:
+    if isinstance(value, str):
+        return Column(glotexp.Literal(this=str(value), is_string=True))
```
exp.Literal.string(...)

tobymao commented on sqlglot/dataframe/sql/functions.py:
```diff
@@ -0,0 +1,1251 @@
+
+
+def grouping_id(*cols: ColumnOrName) -> Column:
+    if len(cols) == 0:
```
not cols

tobymao commented on sqlglot/dataframe/sql/session.py:
```diff
@@ -0,0 +1,145 @@
+        if schema is not None:
+            column_mapping = get_column_mapping_from_schema_input(schema)
+        elif isinstance(data[0], dict):
+            column_mapping = {col_name.strip(): None for col_name in data[0].keys()}
```
you don't need keys

barakalon commented on sqlglot/expressions.py:
```diff
@@ -3131,6 +3134,14 @@ def to_identifier(alias, quoted=None):
     return identifier
 
 
+def to_table(sql_path, pivots=None):
```
I think this is the first time the concept of a "sql_path" has entered this file.

Should this function live here or somewhere else?

eakmanrq replied:
I think it makes sense here alongside the other function to help construct expressions (like `to_alias` and `to_identity`)

barakalon commented on sqlglot/helper.py:
```diff
@@ -24,7 +27,17 @@ def list_get(arr, index):
     if value is None:
         return []
-    return value if isinstance(value, (list, tuple, set)) else [value]
+    return [*value] if isinstance(value, (list, tuple, set)) else [value]
```
I think this introduces loads of copy operations. throughout sqlglot. 

Based on how we use `ensure_list`, I think that's _safe_, but I wonder how it effects performance?

tobymao replied:
agree, we shoudln't do this as it will hurt perf

eakmanrq replied:
Without this change though we aren't really ensuring it is a list which is confusing.

tobymao replied:
it’s just a bad name. it’s ensure enumerable, the better way is to do list(x) where you actually need a list

eakmanrq replied:
Ok I reverted this change

barakalon commented on sqlglot/optimizer/pushdown_projections.py:
```diff
@@ -37,12 +41,16 @@ def pushdown_projections(expression):
 
-        if isinstance(scope.expression, exp.Select):
-            _remove_unused_selections(scope, parent_selections)
+        if isinstance(scope.expression, exp.Select) and scope != right_union:
```
`right_union` can be undefined

eakmanrq replied:
I think line 32 ensures that it will be defined unless I am missing something.

tobymao commented on sqlglot/dataframe/sql/window.py:
```diff
@@ -0,0 +1,116 @@
+        cols = flatten(cols) if isinstance(cols[0], (list, set, tuple)) else cols
+        expressions = [Column.ensure_col(x).expression for x in cols]
+        window_spec = self.copy()
+        if window_spec.expression.args.get("partition_by") is None:
```
probabyl can just do not get partition_by

tobymao commented on sqlglot/dataframe/sql/window.py:
```diff
@@ -0,0 +1,116 @@
+        expressions = [Column.ensure_col(x).expression for x in cols]
+        window_spec = self.copy()
+        if window_spec.expression.args.get("partition_by") is None:
+            window_spec.expression.args["partition_by"] = expressions
```
don't ever set stuff manually, do set or append, here with line 62

tobymao commented on sqlglot/dataframe/sql/window.py:
```diff
@@ -0,0 +1,116 @@
+        expressions = [Column.ensure_col(x).expression for x in cols]
+        window_spec = self.copy()
+        if window_spec.expression.args.get("order") is None:
+            window_spec.expression.args["order"] = exp.Order(expressions=[])
```
same here

barakalon commented on sqlglot/optimizer/pushdown_projections.py:
```diff
@@ -37,12 +41,16 @@ def pushdown_projections(expression):
+            removed_indexes = _remove_unused_selections(scope, parent_selections)
+            # The left union is used for column names to select and if we remove columns from the left
+            # we need to also remove those same columns in the right that were at the same position
+            if scope == left_union:
```
When can this happen?

barakalon replied:
I don't think scope defines an `__eq__` method?

eakmanrq replied:
Good callout. It was falling back to an object instance comparison so I changed this to `is` instead of `==`.

tobymao commented on sqlglot/dialects/spark.py:
```diff
@@ -18,12 +14,6 @@ def _create_sql(self, e):
     return create_with_partitions_sql(self, e)
 
 
-def _map_sql(self, expression):
```
why did you remove this?

eakmanrq replied:
So when I started added in the map functions from the PySpark DataFrame API I noticed that we weren't actually using the map function from Spark but rather creating maps using `MAP_FROM_ARRAYS` even if there weren't arrays. Since Spark had direct support for creating maps like one would expect I removed this workaround.

tobymao replied:
map in spark is Map(a, b, c, d)
in order to do arrays you need to do MAP_FROM_ARRAYS, are you sure this is correct to remove?

tobymao commented on sqlglot/expressions.py:
```diff
@@ -1286,6 +1287,8 @@ def group_by(self, *expressions, append=True, dialect=None, copy=True, **opts):
         Returns:
             Select: the modified expression.
         """
+        if not expressions:
```
how come?

eakmanrq replied:
So the way that the PySpark DataFrame API thinks about this is that an empty group by prior to an aggregate is valid and represents grouping by nothing. This change allows the sqlglot builder to behave in the same way. If you would prefer for the builder to not behave that way then I can remove this and then do a check in the DataFrame API version.

tobymao commented on sqlglot/expressions.py:
```diff
@@ -3131,6 +3134,14 @@ def to_identifier(alias, quoted=None):
     return identifier
 
 
+def to_table(sql_path, pivots=None):
```
docstring

barakalon commented on sqlglot/schema.py:
```diff
@@ -1,13 +1,24 @@
 class Schema(abc.ABC):
     """Abstract base class for database schemas"""
 
+    @abc.abstractmethod
```
Adding an abstractmethod makes this a backwards incompatible change.

I also think of `Schema` as being owned by the optimizer - i.e., it defines the interface that the optimizer depends on.

I think we want a second interface - `MutableSchema`.

tobymao replied:
does this need to be an abc?

eakmanrq replied:
@barakalon Yeah I think the second `MutableSchema` interface makes sense. Working on that change now.

tobymao commented on sqlglot/helper.py:
```diff
@@ -24,7 +27,17 @@ def list_get(arr, index):
     if value is None:
         return []
-    return value if isinstance(value, (list, tuple, set)) else [value]
+    return [*value] if isinstance(value, (list, tuple, set)) else [value]
```
what's this?

eakmanrq replied:
reverted

barakalon commented on sqlglot/schema.py:
```diff
@@ -121,6 +171,24 @@ def ensure_schema(schema):
 
 
+def ensure_column_mapping(mapping):
+    from sqlglot.dataframe.sql import types as df_types
```
This dependency is a smell IMO.

Should the dataframe API have its own schema?

eakmanrq replied:
Yeah going that direction now: https://github.com/tobymao/sqlglot/pull/568#discussion_r992944368

barakalon reviewed (commented):
wowow. This is crazy.

tobymao commented on tests/dialects/test_duckdb.py:
```diff
@@ -92,7 +92,7 @@ def test_duckdb(self):
                 "presto": "SPLIT(x, 'a')",
                 "hive": "SPLIT(x, CONCAT('\\\\Q', 'a'))",
-                "spark": "SPLIT(x, CONCAT('\\\\Q', 'a'))",
+                "spark": "SPLIT(x, 'a')",
```
you tested this right?

tobymao replied:
confirm should revert

tobymao commented on tests/dialects/test_presto.py:
```diff
@@ -40,7 +40,7 @@ def test_cast(self):
                 "presto": "CAST(MAP(ARRAY[1], ARRAY[1]) AS MAP(INTEGER, INTEGER))",
                 "hive": "CAST(MAP(1, 1) AS MAP<INT, INT>)",
-                "spark": "CAST(MAP_FROM_ARRAYS(ARRAY(1), ARRAY(1)) AS MAP<INT, INT>)",
+                "spark": "CAST(MAP(1, 1) AS MAP<INT, INT>)",
```
why did you change this?

eakmanrq replied:
reverted

eakmanrq commented on sqlglot/dataframe/sql/_typing.pyi:
```diff
@@ -0,0 +1,19 @@
+from __future__ import annotations
```
@tobymao and I are discussing now what to do about the types. The reason for including types was that the PySpark DataFrame API has type support and therefore provides really good autocomplete and suggestions to users. I didn't want to have this replacement provide a worse experience for them. I'm open though to switching to docstring types if that is the long term vision for type support in sqlglot.

eakmanrq commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+        return [cls.ensure_col(x) if not isinstance(x, Column) else x for x in args]
+
+    @classmethod
+    def _lit(cls, value: t.Optional[t.Any] = None) -> Column:
```
`lit` is from the PySpark DataFrame API and it stands for `Literal`.

tobymao commented on sqlglot/__init__.py:
```diff
@@ -21,12 +21,15 @@
 
 pretty = False
 
+schema = MappingSchema()
```
i think so, we can discuss later

izeigerman commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+
+
+class Column:
+    def __init__(self, expression: t.Union[t.Any, exp.Expression]):
```
[Nitpick] Seems like we expect very specific types in the constructor. Should we enumerate them in the Union instead of just using Any?

eakmanrq replied:
Yes this was a bad shortcut that I forgot to fix. Will address this as I integrate mypy and clean up the existing types.

izeigerman commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,730 @@
+        subset: t.Optional[t.Union[str, t.Tuple[str, ...], t.List[str]]] = None,
+    ) -> DataFrame:
+        if self.expression.ctes[-1].find(exp.Star) is not None:
+            raise RuntimeError("Cannot use `dropna` when a * expression is used")
```
[Nitpick] we should consider using more specialized error classes

izeigerman commented on sqlglot/schema.py:
```diff
@@ -149,6 +217,35 @@ def _nested_get(d, *path):
+        if key not in subd:
+            subd = subd.setdefault(key, {})
+        else:
+            subd = subd[key]
```
what if the target key is not a dict?

eakmanrq replied:
We know that it will be given the structure of the schema mapping we are using in this file. A good reason why this function exists here and not as a general utility function because of assumptions like that.

izeigerman commented on sqlglot/schema.py:
```diff
@@ -49,37 +60,49 @@ class MappingSchema(Schema):
-
-        self.forbidden_args = {"catalog", "db", "this"} - set(self.supported_table_args)
+        self.supported_table_args = []
+        self.forbidden_args = []
```
Shall we call this `forbidden_table_args` to clarify and to be consistent with `supported_table_args`?

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+        )
+        return Column(new_expression)
+
+    def binary_op(self, clazz: t.Callable, other: Column, **kwargs) -> Column:
```
replace all clazz with klass

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+        column = cls.ensure_col(column) if column is not None else None
+        new_expression = (
+            callable_expression(this=column.column_expression, **kwargs)
+            if column is not None
```
for all of these i prefer the inverse which is

if column is None else ... you have a couple of these calls

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+                )
+            )
+        if isinstance(value, datetime.date):
+            date_literal = exp.Literal(this=value.strftime("%Y-%m-%d"), is_string=True)
```
exp.Literal.string

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+        if isinstance(value, datetime.date):
+            date_literal = exp.Literal(this=value.strftime("%Y-%m-%d"), is_string=True)
+            return cls(
+                exp.Column(this=exp.StrToDate(this=date_literal, format=exp.Literal(this="YYYY-MM-DD", is_string=True)))
```
same

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+
+    @property
+    def alias_or_name(self) -> str:
+        if isinstance(self.expression, exp.Null):
```
can you just do this in base sqlglot?

tobymao commented on sqlglot/dataframe/sql/const.py:
```diff
@@ -0,0 +1,10 @@
+JOIN_HINTS = {
```
can we call it constants.py

tobymao replied:
was a separate file necessary? could have just put it an the top where this was referenced

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,730 @@
+from sqlglot import expressions as exp
+from sqlglot.dataframe.sql import functions as F
+from sqlglot.dataframe.sql.column import Column
+from sqlglot.dataframe.sql import const
```
constants import as c

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,730 @@
+        sequence_id: str = None,
+        last_op: t.Optional[Operation] = Operation.INIT,
+        pending_hints: t.List[exp.Expression] = None,
+        output_expression_container: OutputExpressionContainer = exp.Select(),
```
isn't it dangerous to do this because python evaluates Select() once, it should probably be None and then you do container = container or exp.Select()

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,730 @@
+
+    @classmethod
+    def _create_hash_from_expression(cls, expression: exp.Select):
+        value = expression.sql(dialect="spark").encode("utf-8")
```
does this need to be consistent? we also have _normalize_args

eakmanrq replied:
The goal of this is to produce a consistent hash that is determined by the logic inside the CTE. The reason why we want it consistent is because that way if someone uses the tool to generate SQL to check it into VCS then they SQL will not change if their logic doesn't change.

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,730 @@
+                raise ValueError(f"Invalid expression type: {expression_type}")
+            output_expressions.append(expression)
+
+        return [expression.sql(**{"dialect": dialect, "pretty": True, **kwargs}) for expression in output_expressions]
```
pretty default

eakmanrq replied:
Is the issue that it is pretty by default? I think users of the DataFrame API will want pretty by default. I'm not expecting many use cases where individuals will want the minimal SQL returns for performance reasons.

tobymao replied:
yep, consistency

tobymao replied:
so let's get rid of the pretty default because everywhere else pretty is not default

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,730 @@
+        return [expression.sql(**{"dialect": dialect, "pretty": True, **kwargs}) for expression in output_expressions]
+
+    def copy(self, **kwargs) -> DataFrame:
+        kwargs = {**{k: copy(v) for k, v in vars(self).copy().items()}, **kwargs}
```
i'm not a big fan of this pattern, i prfer things inline

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,730 @@
+    @operation(Operation.SELECT)
+    def distinct(self) -> DataFrame:
+        expression = self.expression.copy()
+        expression.set("distinct", exp.Distinct())
```
make this a sqlglot base method

eakmanrq replied:
I'm not sure what you mean by this.

tobymao replied:
make a method in expression.py that does this

eakmanrq replied:
Ah I see there already is one. Will switch to that.

tobymao commented on sqlglot/dataframe/sql/readwriter.py:
```diff
@@ -0,0 +1,76 @@
+        self._by_name = by_name
+
+    def copy(self, **kwargs) -> DataFrameWriter:
+        kwargs = {**{k: copy(v) for k, v in vars(self).copy().items()}, **kwargs}
```
second time i've seen this, let's make a helper

izeigerman commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+            expression = sqlglot.parse_one(expression, read="spark")
+        elif isinstance(expression, Column):
+            expression = expression.expression
+        self.expression = expression
```
Do we need to ensure the validity of the passed expression in any way? I assume not every expression can be `Column`.

eakmanrq replied:
Many can. I think I'm actually good with this approach for now and see if we get any feedback about issues. My guess is that this won't be an issue.

izeigerman commented on sqlglot/dataframe/sql/session.py:
```diff
@@ -0,0 +1,145 @@
+        from sqlglot.dataframe.sql.dataframe import DataFrame
+
+        if samplingRatio is not None or verifySchema:
+            raise NotImplementedError("Sampling Ration and Verify Schema are not supported")
```
Typo in my Type lol

eakmanrq commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,730 @@
+    @operation(Operation.FROM)
+    def dropna(
+        self,
+        how: str = "any",
```
This line covers the all case: `minimum_num_nulls = 1 if how == "any" else len(null_check_columns)`

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,320 @@
+        elif isinstance(expression, Column):
+            expression = expression.expression
+        self.expression = expression
+        if self.expression is None or isinstance(self.expression, (int, float, bool, dict, Iterable, datetime.date)):
```
would it be better to do like

if not isinstance(expression, exp.Expression):

eakmanrq replied:
Yeah good call updated

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,320 @@
+        return isinstance(self.expression, exp.Column)
+
+    @property
+    def column_expression(self) -> exp.Column:
```
this is a duplicate of expression.unalias()

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,320 @@
+        expression.set("table", exp.to_identifier(table_name))
+        return Column(expression)
+
+    def sql(self, **kwargs) -> Column:
```
allow to pass in dialect too?

sql(**{"dialect": "spark", **kwargs})

eakmanrq replied:
awesome great addition

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,320 @@
+        new_expression = exp.Not(this=exp.Is(this=self.column_expression, expression=exp.Null()))
+        return Column(new_expression)
+
+    def cast(self, dataType: t.Union[str, "DataType"]):
```
use the real DataType?

tobymao commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,320 @@
+            self, exp.Substring, start=startPos.expression, length=length.expression
+        )
+
+    def isin(self, *cols: t.Union["Primitives", t.Iterable["Primitives"]]):
```
types in this file need to be updated

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,737 @@
+        branch_id: str = None,
+        sequence_id: str = None,
+        last_op: t.Optional[Operation] = Operation.INIT,
+        pending_hints: t.List[exp.Expression] = None,
```
needs to be t.Optional

tobymao commented on tests/test_executor.py:
```diff
@@ -7,11 +8,13 @@
 DIR = FIXTURES_DIR + "/optimizer/tpc-h/"
 
 
+@unittest.skipIf(string_to_bool(os.environ.get("SKIP_INTEGRATION", '0').lower()),
```
can you make this var reusable?

tobymao commented on tests/dataframe/integration/dataframe_validator.py:
```diff
@@ -0,0 +1,151 @@
+    from pyspark.sql import DataFrame as SparkDataFrame
+
+
+@unittest.skipIf(string_to_bool(os.environ.get("SKIP_INTEGRATION", '0').lower()),
```
yea lets just import SKIP_INTEGRATION instead of this whole thing

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,737 @@
+                    matching_ctes = [
+                        cte for cte in reversed(expression.ctes) if cte.args["sequence_id"] in sequence_ids_to_match
+                    ]
+                    if matching_ctes:
```
the if statement is unnecessary

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,737 @@
+                    ]
+                    if matching_ctes:
+                        for matching_cte in matching_ctes:
+                            if matching_cte.alias_or_name in [join_table.alias_or_name for join_table in join_tables]:
```
take this for loop outside and make it a set?

join_aliases = {join_table.alias_or_name for . in .}

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,737 @@
+        new_df.pending_hints.append(hint_expression)
+        return new_df
+
+    def _set_operation(self, clazz: t.Callable, other: DataFrame, distinct: bool):
```
change all clazz to klass

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,737 @@
+        expression = expression.copy()
+        with_expression = expression.args.get("with")
+        if with_expression:
+            existing_ctes = with_expression.args["expressions"]
```
with_expression.expressions

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,737 @@
+        with_expression = expression.args.get("with")
+        if with_expression:
+            existing_ctes = with_expression.args["expressions"]
+            existsing_cte_names = [x.alias_or_name for x in existing_ctes]
```
make this a set {name for .. .}

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,737 @@
+    @classmethod
+    def _get_outer_select_columns(cls, item: t.Union[exp.Expression, DataFrame]) -> t.List[Column]:
+        expression = item.expression if isinstance(item, DataFrame) else item
+        return [Column(x) for x in dict.fromkeys(expression.find(exp.Select).args.get("expressions", []))]
```
are you trying to deduplicate this?

tobymao replied:
find(exp.Select).expressions, automatically gives you the empty list default

eakmanrq replied:
Yeah I think this can actually be removed. Thanks for calling it out.

tobymao commented on sqlglot/dataframe/sql/dataframe.py:
```diff
@@ -0,0 +1,737 @@
+        all_column_mapping = {column.alias_or_name: column for column in all_columns}
+        if isinstance(value, dict):
+            values = value.values()
+            columns = self._ensure_and_normalize_cols(list(value.keys()))
```
list(value) don't need keys

tobymao commented on sqlglot/dataframe/sql/session.py:
```diff
@@ -0,0 +1,145 @@
+
+
+class SparkSession:
+    known_ids: t.ClassVar[t.Set[str]] = set()
```
mutable class vars? this is intended?

eakmanrq replied:
Yes and I believe this will be fine in practice. I think technically though this could be a singleton if we wanted.

tobymao commented on sqlglot/dialects/spark.py:
```diff
@@ -98,20 +88,22 @@ class Generator(Hive.Generator):
 
         TRANSFORMS = {
-            **{k: v for k, v in Hive.Generator.TRANSFORMS.items() if k not in {exp.ArraySort}},
+            **{k: v for k, v in Hive.Generator.TRANSFORMS.items() if k not in {exp.ArraySort, exp.Split, exp.ILike}},
```
remove the split?

eakmanrq replied:
Hmm I did I think your page hadn't refreshed.

tobymao commented on sqlglot/schema.py:
```diff
@@ -49,37 +60,49 @@ class MappingSchema(Schema):
+        return ("this",)
+
+    def copy(self, **kwargs):
+        kwargs = {**{"schema": copy(self.schema)}, **kwargs}
```
why not just do self.schema.copy()

tobymao replied:
why is this double nested?

MappingSchema(**{"schema": self.schema.copy(), **kwargs})

eakmanrq commented on sqlglot/dataframe/sql/column.py:
```diff
@@ -0,0 +1,330 @@
+
+    @classmethod
+    def _lit(cls, value: t.Optional[t.Any] = None) -> Column:
+        if value is None:
```
Great suggestion. I wish I had known about that converter earlier. The only difference was how we handled dictionaries so I left that part separate.

tobymao commented on sqlglot/expressions.py:
```diff
@@ -3321,6 +3346,12 @@ def convert(value):
             values=[convert(v) for v in value.values()],
         )
+    if isinstance(value, datetime.datetime):
+        datetime_literal = Literal.string(value.strftime("%Y-%m-%d %H:%M:%S"))
```
i don't believe these are correct. you're hard-coding hive date formats here. it might be more generic to do

Cast(this=literal, to=DataType.build("timestamp"))

Also should you handle timezones? in that case maybe you want utc time or using iso format

this holds the same for date

i also have a method called TimeStringToTime and DateStringToDate, the issue with those functions is that i haven't implemented them for every dialect yet, although we could make those default to CAST(...)

this may be the way to go

tobymao commented:
you added a bad file test_dialects.py

eakmanrq commented on sqlglot/dataframe/sql/functions.py:
```diff
@@ -0,0 +1,1251 @@
+    return Column(value)
+
+
+def greatest(*cols: ColumnOrName) -> Column:
```
I don't want to hold up this PR for that feature so I created an issue: https://github.com/tobymao/sqlglot/issues/597

eakmanrq commented:
Addressed all the feedback and added mypy. I did add some ignores but not many and one of the main issues was around dealing with inputs that may or may not be nested and therefore I may or may not flatten.

barakalon commented on sqlglot/schema.py:
```diff
@@ -0,0 +1,297 @@
+
+class MutableSchema(Schema):
+    """
+    Based on MappingSchema but mutable so you can add tables over time.
```
Strange that the subclass is "based on the MappingSchema", a superclass.

I think this class should be a lot simpler. It should be more of an interface.

tobymao replied:
i would have expected the default Mapping Schema to come first

eakmanrq replied:
I think that docstring was just not updated after the refactor. 

Mapping schema did come first before the last request for changes. The last request asked for this hierarchy of Schema -> MutableSchema(Schema) -> MappingSchema(MutableSchema)

barakalon replied:
Right now, all `MappingSchema` does is _unimplement_ add_table. That's silly.

I think it's fine for the the MappingSchema to implement add_table. I just didn't think add_table should be part of the main Schema interface.

We only need this intermediate interface, MutableSchema, if the DataFrame API wants to support custom Schema types.

barakalon commented on sqlglot/schema.py:
```diff
@@ -0,0 +1,300 @@
+    @abc.abstractmethod
+    def add_table(self, table, column_mapping=None):
+        """
+        Registers the table to be a known schema to be accessed later. Some implementing classes may
```
Can you rephrase this description? It's not very clear.
